### ISSUE TITLE
Replace "x.has_key(y)" with "y in x" where covered by tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5688]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)| @vperic reported|
|-|-|
|Trac ID|trac#5688|
|Type|task|
|Created|2012-05-28 13:05:19Z|
|Branch|https://github.com/twisted/twisted/tree/haskey-removal-5688|

The dict.has_key(k) syntax is deprecated (and completely gone by Python 3) in favor of "k in dict". Ticket [#4053](https://github.com/twisted/twisted/issues/4053) attempted to fix this, but was stalled due to making many untested changes (though these are all trivial). As such, I've decided to take a different approach: this ticket only fixes the lines covered by the tests suite. I determined this by running the test suite with the "-3" option, and then checking each line individually.

Attachments:

* [haskey-tested.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/357/3578d6baffece6f779e52dcf1ff9625e0c8ad443/ad637f6fed533681b3cac362d26ee568a9ed1462.patch) (21233 bytes) - added by vperic on 2012-05-28 13:05:42Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5688 5688
type__task task
reporter__vperic vperic
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_haskey_removal_5688 branches/haskey-removal-5688
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__py3k py3k
time__1338210319000000 1338210319000000
changetime__1339074943000000 1339074943000000
version__None None
owner__vperic vperic

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst set owner to @vperic|
|-|-|

Thanks for working on this! Sticking to tested stuff in one ticket and untested stuff in another ticket seems like a reasonable approach. A few more minor changes needed:

  1. Update coding standard document to note that has_key shouldn't be used.
  1. Add a .misc newsfile.
  1. Once ticket is merged, add a comment to [#4053](https://github.com/twisted/twisted/issues/4053) indicating its scope is now limited to instances that are untested.

Next steps:
  1. Do the above changes in a Twisted branch using the combinator infrastructure, mkbranch and so on (I'll make sure you get commit access).
  1. Run force-builds.py to make sure it passes on all platforms (http://twistedmatrix.com/trac/wiki/CommitterCheckList has a link). There may BTW be e.g. Windows-only tests that need has_key fixed as well, but those can be done elsewhere in another ticket, don't have to do all at once.
  1. Merge.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_2"></a>|@vperic commented|
|-|-|

(In [34486]) Create branch haskey-removal-5688

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_3"></a>|@vperic commented|
|-|-|

Ok, so, assuming I did everything correctly, the branch is ok (with a topfile but without coding standard changes; I'll compile several important points and submit a new ticket) and I'm the builds should be at: [Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/haskey-removal-5688)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

If you meant to submit this for review again you should add "review" keyword.

You're still missing the news file, I think? Once that is in merge.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_5"></a>|@vperic removed owner|
|-|-|

Ok, added the topfile, run the tests again, but there are still failures in three of the bots. I don't really see code of mine could have caused any of this, but as this is my first time merging, I'd still prefer to have someone look over it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set owner to @vperic|
|-|-|

Those look like the usual failures (not sure what's up with lucid buildslave, but that also looks unrelated to you). So looks ready for merging.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_7"></a>|@vperic set status to `closed`|
|-|-|

(In [34521]) Merge haskey-removal-5688

Author: vperic
Reviewer: itamar
Fixes: [#5688](https://github.com/twisted/twisted/issues/5688)

The "dict.has_key(k)" syntax is deprecated in Python 3, use the new
"k in dict" syntax instead. This change only covers the uses of has_key
covered by the tests.

In one case, the has_key function itself was being passed; pass the
internal __contains__ function instead.
